etcdadm/etcdadm

```diff
@@ -0,0 +1,1621 @@
+        # the snapshot contained the information about this member hence it is recognized to be "unstarted" by other members,
+        # instead of just being invisible from them.
+        # In other words, the cluster is still in process of a disaster recovery and this is the `N/2+1`th or later member.
+        _info 'cluster is already healthy but still in bootstrap process after the disaster recovery. searching for a etcd snapshot to recover this member'
+        member_bootstrap
+      fi
+    elif member_is_failing_longer_than_limit; then
+      # This member seems to be consistently failing
+      #
+      # As the cluster is still healthy, it can happen only when:
+      # * the etcd data of this member is broken somehow or
+      # * this member has a network connectivity issue between other members in the cluster
+      # The latter should be eventually managed by operators or AWS.
+      # To deal with the former case, we just restart this member with fresh data.
+      #
+      # This process is documented in the section "Replace failed etcd member" in the etcd documentation.
+      # See https://coreos.com/etcd/docs/latest/etcd-live-cluster-reconfiguration.html#replace-a-failed-etcd-member-on-coreos-container-linux
+      _info 'this member is failing longer than limit'
+      member_replace_failed
+    else
+      # This member has just been restarted.
+      #
+      # The restart may have been caused by the following reasons:
+      # * EC2 instance which had been hosting this member terminated due to a failure, and then the ASG recreated it
+      # * The user initiated a reboot of the EC2 instance hosting this member
+      # Although there's no way to certainly determine which one it is,
+      # we can safely retry until the failing period exceeds the threshold and hope the member eventually becomes healthy
+      # if the failure is'nt permanent.
+      _info 'this member has just restarted'
+    fi
+  else
+    # At least N/2+1 members are NOT working
+
+    local running_num
+    local remaining_num
+    local total_num
+    running_num=$(cluster_num_running_nodes)
+    total_num=$member_count
+    remaining_num=$(( quorum - running_num + 1 ))
+
+    _info "${remaining_num} more nodes are required until the quorum is met"
+
+    if (( remaining_num >= 2 )); then
+      member_set_unit_type simple
+    else
+      member_set_unit_type notify
+    fi
+
+    if (( running_num < total_num )); then
+      _info "only ${running_num} of ${total_num} nodes for etcd members are running, which means cluster is still in bootstrap process. searching for a etcd snapshot to recover this member"
+      member_bootstrap
+    elif cluster_is_failing_longer_than_limit; then
+      _info "all the nodes for etcd members are running but cluster has been unhealthy for a while, which means cluster is now in disaster recovery process. searching for a etcd snapshot to recover this member"
+      member_bootstrap
+    else
```

Comment:
Perhaps this block results in the last etcd node to miss restoring itself from an etcd snapshot, if any?

We can probably change this if-else block to just:

```bash
f (( running_num < total_num )); then
  _info "only ${running_num} of ${total_num} nodes for etcd members are running, which means cluster is still in bootstrap process. searching for a etcd snapshot to recover this member"
elif (( running_num == total_num )); then
  _info "all the ${running_num} of ${total_num} nodes for etcd members are running. Almost certainly this node is the last one?"
elif cluster_is_failing_longer_than_limit; then
  _info "all the nodes for etcd members are running but cluster has been unhealthy for a while, which means cluster is now in disaster recovery process. searching for a etcd snapshot to recover this member"
else
  _panic "[bug] invalid state!"
fi
member_bootstrap
```